src/components/implementation/no_interface/vkernel/vkernel.c

```diff
@@ -0,0 +1,195 @@
+
+		printc("\tCopying pgtbl, captbl, component capabilities\n");
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_CT, vk_cinfo, vm_info->ct);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_PT, vk_cinfo, vm_info->pt);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_UNTYPED_PT, vk_cinfo, vm_info->utpt);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_COMP, vk_cinfo, vm_info->cc);
+		assert(ret == 0);
+
+		printc("\tCreating and copying required initial capabilities\n");
+		/*
+		 * TODO: Multi-core support to create INITIAL Capabilities per core
+		 */
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITTHD_BASE, vk_cinfo, vm_info->initthd);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITHW_BASE, vk_cinfo, BOOT_CAPTBL_SELF_INITHW_BASE);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, VM_CAPTBL_SELF_EXITTHD_BASE, vk_cinfo, vm_info->exitthd);
+		assert(ret == 0);
+		
+		vm_info->inittcap = cos_tcap_alloc(vk_cinfo, TCAP_PRIO_MAX);
+		assert(vm_info->inittcap);
+
+		vm_info->initrcv = cos_arcv_alloc(vk_cinfo, vm_info->initthd, vm_info->inittcap, vk_cinfo->comp_cap, BOOT_CAPTBL_SELF_INITRCV_BASE);
+		assert(vm_info->initrcv);
+
+		ret = cos_tcap_transfer(vm_info->initrcv, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, TCAP_PRIO_MAX);
+		assert(ret == 0);
+
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITTCAP_BASE, vk_cinfo, vm_info->inittcap);
+		assert(ret == 0);
+		ret = cos_cap_cpy_at(vm_cinfo, BOOT_CAPTBL_SELF_INITRCV_BASE, vk_cinfo, vm_info->initrcv);
+		assert(ret == 0);
+
+		/*
+		 * Create send end-point in VKernel to each VM's INITRCV end-point
+		 */
+		vk_info.vminitasnd[id] = cos_asnd_alloc(vk_cinfo, vm_info->initrcv, vk_cinfo->captbl_cap);
+		assert(vk_info.vminitasnd[id]);
+
+		/*
+		 * Create and copy booter comp virtual memory to each VM
+		 */
+		vm_range = (int)cos_get_heap_ptr() - BOOT_MEM_VM_BASE;
+		assert(vm_range > 0);
+		printc("\tMapping in Booter component's virtual memory (range:%u)\n", vm_range);
+		for (i = 0; i < vm_range; i += PAGE_SIZE) {
+			vaddr_t spg = (vaddr_t)cos_page_bump_alloc(vk_cinfo), dpg;
+			assert(spg);
+			
+			memcpy((void *)spg, (void *)(BOOT_MEM_VM_BASE + i), PAGE_SIZE);
+			
+			dpg = cos_mem_alias(vm_cinfo, vk_cinfo, spg);
+			assert(dpg);
+		}
+
+		printc("\tAllocating Untyped memory\n");
+		cos_meminfo_alloc(vm_cinfo, BOOT_MEM_KM_BASE, COS_VIRT_MACH_UNTYPED_SIZE);
+
+		printc("vkernel: VM%d Init END\n", id);
+	}
+
+	printc("Starting Scheduler\n");
+	cos_hw_attach(BOOT_CAPTBL_SELF_INITHW_BASE, HW_PERIODIC, BOOT_CAPTBL_SELF_INITRCV_BASE);
+	printc("\t%d cycles per microsecond\n", cos_hw_cycles_per_usec(BOOT_CAPTBL_SELF_INITHW_BASE));
+	printc("------------------[ VKernel & VMs init complete ]------------------\n");
+
+	scheduler();
+
+	cos_hw_detach(BOOT_CAPTBL_SELF_INITHW_BASE, HW_PERIODIC);
+
+	printc("vkernel: END\n");
+	cos_thd_switch(vk_info.termthd);
+
+	printc("DEAD END\n");
```

Comment:
It is kind of an unreachable code and the error-report I had, yeah. There is probably no useful information I could print here as part of error report.